File: src/notify/notifier.ts
import * as core from "@actions/core";
import type { Config } from "../config.js";
import type { Audit } from "../util/audit.js";
import type { NotifyChannel, Report } from "./types.js";
import { SlackChannel } from "./slack.js";
import { EmailChannel } from "./email.js";

/**
 * Fans a {@link Report} out to every configured external channel. Records each
 * delivery in the audit trail, honors `dryRun`, and never lets one channel's
 * failure abort the run (or the other channels).
 */
export class Notifier {
  constructor(
    private readonly channels: NotifyChannel[],
    private readonly dryRun: boolean,
    private readonly audit: Audit,
  ) {}

  /** Number of active channels (0 means notifications are effectively off). */
  get channelCount(): number {
    return this.channels.length;
  }

  async broadcast(report: Report): Promise<void> {
    for (const ch of this.channels) {
      this.audit.record(report.feature, `notify:${ch.name}`, ch.target, report.title);
      if (this.dryRun) continue;
      try {
        await ch.send(report);
      } catch (err) {
        core.warning(`Notification via ${ch.name} failed: ${(err as Error).message}`);
      }
    }
  }
}

/**
 * Build a Notifier from config. Secrets (webhook URL, SMTP credentials) are read
 * from environment variables named in the config — never from the config file
 * itself — so they can be supplied as encrypted CI secrets.
 */
export function buildNotifier(cfg: Config, env: NodeJS.ProcessEnv, dryRun: boolean, audit: Audit): Notifier {
  const channels: NotifyChannel[] = [];
  const n = cfg.notifications;

  if (n?.slack?.enabled) {
    const url = env[n.slack.webhookEnv];
    if (!url) {
      core.warning(`Slack notifications enabled but env var "${n.slack.webhookEnv}" is empty; skipping Slack.`);
    } else {
      channels.push(new SlackChannel(url));
    }
  }

  if (n?.email?.enabled) {
    channels.push(
      new EmailChannel({
        host: n.email.host,
        port: n.email.port,
        secure: n.email.secure,
        user: n.email.userEnv ? env[n.email.userEnv] : undefined,
        pass: n.email.passwordEnv ? env[n.email.passwordEnv] : undefined,
        from: n.email.from,
        to: n.email.to,
      }),
    );
  }

  return new Notifier(channels, dryRun, audit);
}

File: ruby/lib/boardly/notify/notifier.rb
# frozen_string_literal: true

require_relative "slack"

module Boardly
  module Notify
    # Fans a Report out to every configured external channel. Records each
    # delivery in the audit trail, honors dry_run, and never lets one channel's
    # failure abort the run (or the other channels).
    class Notifier
      def initialize(channels, dry_run, audit)
        @channels = channels
        @dry_run = dry_run
        @audit = audit
      end

      def channel_count = @channels.length

      def broadcast(report)
        @channels.each do |ch|
          @audit.record(report.feature, "notify:#{ch.name}", ch.target, report.title)
          next if @dry_run

          begin
            ch.send(report)
          rescue StandardError => e
            puts "::warning::Notification via #{ch.name} failed: #{e.message}"
          end
        end
      end

      # Build a Notifier from config. Secrets (webhook URL, SMTP credentials) are
      # read from ENV vars named in the config — never from the config file.
      def self.build(cfg, env, dry_run, audit)
        channels = []
        n = cfg.notifications || {}

        if n.dig(:slack, :enabled)
          url = env[n.dig(:slack, :webhook_env)]
          if url.nil? || url.empty?
            puts %(::warning::Slack notifications enabled but env var "#{n.dig(:slack, :webhook_env)}" is empty; skipping Slack.)
          else
            channels << SlackChannel.new(url)
          end
        end

        if n.dig(:email, :enabled)
          require_relative "email"
          e = n[:email]
          channels << EmailChannel.new(
            host: e[:host], port: e[:port], secure: e[:secure],
            user: e[:user_env] && env[e[:user_env]], pass: e[:password_env] && env[e[:password_env]],
            from: e[:from], to: e[:to]
          )
        end

        new(channels, dry_run, audit)
      end
    end
  end
end
